Cython/Compiler/Nodes.py

```diff
@@ -3018,21 +3041,49 @@ def analyse_signature(self, env):
             sig.is_staticmethod = True
             sig.has_generic_args = True
 
-        if ((self.is_classmethod or self.is_staticmethod) and
-                self.has_fused_arguments and env.is_c_class_scope):
-            del self.decorator_indirection.stats[:]
-
         for i in range(min(nfixed, len(self.args))):
             arg = self.args[i]
             arg.is_generic = 0
-            if sig.is_self_arg(i) and not self.is_staticmethod:
+            if (sig.is_self_arg(i) and not self.is_staticmethod):
+                self_type_overridden = arg.type_set_manually
                 if self.is_classmethod:
                     arg.is_type_arg = 1
-                    arg.hdr_type = arg.type = Builtin.type_type
+                    usual_type = Builtin.type_type
+                    if not arg.type_set_manually:
+                        arg.hdr_type = arg.type = usual_type
                 else:
                     arg.is_self_arg = 1
-                    arg.hdr_type = arg.type = env.parent_type
+                    if self.is_in_arbitrary_decorator:
+                        usual_type = PyrexTypes.FusedType([env.parent_type, PyrexTypes.py_object_type],
```

Comment:
Phew – I'm not sure if it's always a good idea to change a "normal" function into a fused one. That's seems bit too non-obvious for users, also regarding performance characteristics etc.

Reply:
I do see the point here. This seemed less painful than a custom dispatch mechanism just for this use.

A few possible options:
* I already added a warning message for arbitrary decorators without an explicit type (https://github.com/cython/cython/pull/3383/files#r405058184). It could be more detailed about what it's done and mention the fused function.
* Compiler option to control this (so keep current behaviour as the default, but allow this extension with a compiler options)
* It might be possible to make a simpler dispatch mechanism - just duplicate the tree inside the function so it's effectively in an `if statement`. Obviously that might well have its own complications:

       ```
       if isinstance(self, CdefClassType):
           # code
       else
            # copy of code with "self" -> "<object>self"
     ```